walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,107 @@
+    Collection<String> attendees = new ArrayList<String>(request.getAttendees());
+    Collection<String> optionalAttendees = new ArrayList<String>(request.getOptionalAttendees());
+    long duration = request.getDuration();
+
+    // if duration is longer than the entire day, no meeting can occur
+    if (duration > TimeRange.WHOLE_DAY.duration()) {
+      return Arrays.asList();
+    }
+    //no attendees or optional attendees
+    if (attendees.size() == 0 && optionalAttendees.size() == 0) {
+      return (Arrays.asList(TimeRange.WHOLE_DAY));
+    }
+    // only mandatory attendees
+    if (attendees.size() > 0 && optionalAttendees.size() ==0){
+      return findTimes(attendees, events, duration);
+    }
+    // only optional attendees
+    if (attendees.size() ==0 && optionalAttendees.size() > 0){
+      return findTimes(optionalAttendees, events, duration);
+    }
+    // both optional and mandatory attendees, try merging the two attendee lists together and finding meeting times.
+    if (attendees.size() > 0 && optionalAttendees.size() >0){
+      attendees.addAll(optionalAttendees);
+      Collection<TimeRange> toReturn = findTimes(attendees, events, duration);
+      if (toReturn.size() > 0) {
+        return toReturn;
+      }
+    }
+    // if no meeting time exists such that mandatory and optional attendees can all go, just find meeting times for mandatory attendees
+    attendees.removeAll(optionalAttendees);
+    return findTimes(attendees, events, duration);
+  }
+
+  //given a list of attendees and events, find meeting times that work
+  public Collection<TimeRange> findTimes(Collection<String> attendees, Collection<Event> events, long duration) {
+    List<TimeRange> conflicts = new ArrayList<TimeRange>();
+    Collection<TimeRange> toReturn = new ArrayList<TimeRange>();
+ 
+    for (Event event : events) {
+      if (containsAttendees(attendees, event.getAttendees())) {
+        conflicts.add(event.getWhen());
+      }
+    }
+
+    if (conflicts.size() >0) {
+      toReturn = conjoinTimes(conflicts, (int) duration);
+      return (toReturn);
+    }
+
+    return (Arrays.asList(TimeRange.WHOLE_DAY));
   }
+
+  //given a list of conflicts, invert the conflict times to find available times
+  public Collection<TimeRange> conjoinTimes(List<TimeRange> conflicts, int duration) {
+    Collection<TimeRange> toReturn = new ArrayList<TimeRange>();
+    int start = TimeRange.START_OF_DAY;
+    int end = TimeRange.END_OF_DAY;
+    Collections.sort(conflicts, TimeRange.ORDER_BY_START);
+    for (int i = 0; i < conflicts.size(); i++) {
+      TimeRange conflict = conflicts.get(i);
+ 
+      if (conflict.start() > start && conflict.start() - start >= duration && start + conflict.start() <= end) {
+        TimeRange toAdd = TimeRange.fromStartDuration(start, conflict.start()-start);
+        toReturn.add(toAdd);
+        start = conflict.end();
+      }
+      else if (conflict.start() <= start && conflict.end() < start) {
+        break;
```

Comment:
I think this `break` may be problematic, should it be `continue`? e.g. given

```
START_OF_DAY = 9
END_OF_DAY = 17

conflicts = [
  0: (10, 13),
  1: (11, 12),
  2: (14, 17),
]

duration = 2
```

(i.e. the second conflict is entirely contained within the first conflict, and there are no valid times)

```
i=0:
  end of loop updates start = 13
i=1:
  hits break because conflict.start and conflict.end are both < start
after loop:
  start is still 13, and so the time range (13, 17) will be returned

If I'm correct, probably should add this as a test case :)